state/pruner.go

```diff
@@ -0,0 +1,343 @@
+	p.mtx.Lock()
+	defer p.mtx.Unlock()
+
+	if height <= 0 || height > p.bs.Height() {
+		return ErrInvalidHeightValue
+	}
+	currentRetainHeight, err := p.stateStore.GetABCIResRetainHeight()
+	if err != nil {
+		if !errors.Is(err, ErrKeyNotFound) {
+			return err
+		}
+		return p.stateStore.SaveABCIResRetainHeight(height)
+	}
+	if currentRetainHeight > height {
+		return ErrPrunerCannotLowerRetainHeight
+	}
+	return p.stateStore.SaveABCIResRetainHeight(height)
+}
+
+// GetApplicationRetainHeight is a convenience method for accessing the
+// GetApplicationRetainHeight method of the underlying state store.
+func (p *Pruner) GetApplicationRetainHeight() (int64, error) {
+	return p.stateStore.GetApplicationRetainHeight()
+}
+
+// GetCompanionBlockRetainHeight is a convenience method for accessing the
+// GetCompanionBlockRetainHeight method of the underlying state store.
+func (p *Pruner) GetCompanionBlockRetainHeight() (int64, error) {
+	return p.stateStore.GetCompanionBlockRetainHeight()
+}
+
+// GetABCIResRetainHeight is a convenience method for accessing the
+// GetABCIResRetainHeight method of the underlying state store.
+func (p *Pruner) GetABCIResRetainHeight() (int64, error) {
+	return p.stateStore.GetABCIResRetainHeight()
+}
+
+func (p *Pruner) pruneABCIResRoutine() {
+	p.logger.Info("Started pruning ABCI results", "interval", p.interval.String())
+	lastABCIResRetainHeight := int64(0)
+	for {
+		select {
+		case <-p.Quit():
+			return
+		default:
+			newABCIResRetainHeight := p.pruneABCIResToRetainHeight(lastABCIResRetainHeight)
+			p.observer.PrunerPrunedABCIRes(&ABCIResponsesPrunedInfo{
+				FromHeight: lastABCIResRetainHeight,
+				ToHeight:   newABCIResRetainHeight - 1,
+			})
+			lastABCIResRetainHeight = newABCIResRetainHeight
+			time.Sleep(p.interval)
+		}
+	}
+}
+func (p *Pruner) pruneBlocksRoutine() {
+	p.logger.Info("Started pruning blocks", "interval", p.interval.String())
+	lastRetainHeight := int64(0)
+	for {
+		select {
+		case <-p.Quit():
+			return
+		default:
+			newRetainHeight := p.pruneBlocksToRetainHeight(lastRetainHeight)
+			p.observer.PrunerPrunedBlocks(&BlocksPrunedInfo{
+				FromHeight: lastRetainHeight,
+				ToHeight:   newRetainHeight - 1,
+			})
+			lastRetainHeight = newRetainHeight
+			time.Sleep(p.interval)
+		}
+	}
+}
+
+func (p *Pruner) pruneBlocksToRetainHeight(lastRetainHeight int64) int64 {
+	targetRetainHeight := p.findMinRetainHeight()
+	if targetRetainHeight == lastRetainHeight {
+		return lastRetainHeight
+	}
+	pruned, evRetainHeight, err := p.pruneBlocks(targetRetainHeight)
+	// The new retain height is the current lowest point of the block store
+	// indicated by Base()
+	newRetainHeight := p.bs.Base()
+	if err != nil {
+		p.logger.Error("Failed to prune blocks", "err", err, "targetRetainHeight", targetRetainHeight, "newRetainHeight", newRetainHeight)
+	} else if pruned > 0 {
+		p.logger.Info("Pruned blocks", "count", pruned, "evidenceRetainHeight", evRetainHeight, "newRetainHeight", newRetainHeight)
```

Comment:
I suppose that, if we're exposing metrics for operators, we could downgrade this to `Debug` level. As long as there's some way of operators knowing whether it's actually working.

Reply:
> way of operators knowing whether it's actually working

Then maybe we could just Info-log every N (e.g., N=500) blocks to reduce verbosity, the rest being Debug-log. I'm convinced that the current verbosity will add noise to the Info logs.

I think we're using (something similar to) this technique elsewhere in this PR.

Reply:
We have metrics now so I think it's all good. Unless you think there is  still a point in logging this ever X blocks in case operators do not use prometheus to monitor the node.

Reply:
With metrics, I think turning this into a Debug should be good enough